src/x/server/secured_conn.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package server implements a network server.
+package server
+
+import (
+	"bufio"
+	"crypto/tls"
+	"net"
+)
+
+// TLSHandshakeFirstByte is the first byte of a tls connection handshake
+const TLSHandshakeFirstByte = 0x16
+
+func newSecuredConn(conn net.Conn) SecuredConn {
+	return &securedConn{
+		r:     bufio.NewReader(conn),
```

Comment:
Instead of using a bufio.Reader for the entirety of the connection, could we just implement peeking of the first byte ourselves? We only need the peek functionality once, and after that we could just read as normal.

My thinking here is that it would save us some unnecessary buffering overhead. May not be that significant of a performance implication, but then again, it's also probably not too complex to just buffer the first byte ourselves.

Reply:
It's done, bufio.Reader was replaced with the peek function.